langchain/vectorstores/redis.py

```diff
@@ -1,128 +1,308 @@
+            from redis.commands.search.query import Query
+        except ImportError:
+            raise ValueError(
+                "Could not import redis python package. "
+                "Please install it with `pip install redis`."
+            )
+
+        # Setup Redis query
+        return_fields = ["metadata", "content", "vector_score"]
+        vector_field = "content_vector"
+        hybrid_fields = "*"
+        base_query = (
+            f"{hybrid_fields}=>[KNN {k} @{vector_field} $vector AS vector_score]"
+        )
+        redis_query = (
+            Query(base_query)
+            .return_fields(*return_fields)
+            .sort_by("vector_score")
+            .paging(0, k)
+            .dialect(2)
+        )
+        params_dict: Mapping[str, str] = {
+            "vector": np.array(embedding)  # type: ignore
+            .astype(dtype=np.float32)
+            .tobytes()
+        }
 
-        self.client = redis_client
+        # Perform vector search
+        results = (
+            await self.client
+              .ft(self.index_name)
+              .search(redis_query, params_dict)
+        )
+
+        # Aggregate result documents
+        return [
+            (
+                Document(
+                    page_content=result.content,
+                    metadata=json.loads(result.metadata)
+                ),
+                float(result.vector_score),
+            )
+            for result in results.docs
+        ]
 
     def add_texts(
         self,
         texts: Iterable[str],
         metadatas: Optional[List[dict]] = None,
+        embeddings: Optional[List[float]] = None,
         **kwargs: Any,
     ) -> List[str]:
         """Add texts data to an existing index."""
-        prefix = _redis_prefix(self.index_name)
+        return self.loop.run_until_complete(
+            self.aadd_texts(texts, metadatas, embeddings, **kwargs)
+        )
+
+    async def aadd_texts(
+        self,
+        texts: Iterable[str],
+        metadatas: Optional[List[dict]] = None,
+        embeddings: Optional[List[float]] = None,
+        **kwargs: Any
+    ) -> List[str]:
+        """Add texts data to an existing index."""
+        prefix = self._redis_prefix(self.index_name)
         keys = kwargs.get("keys")
-        ids = []
-        # Write data to redis
-        pipeline = self.client.pipeline(transaction=False)
+        # Init list of keys
+        ids: List[str] = []
+        # Write data to redis in a pipeline
+        pipeline = await self.client.pipeline(transaction=False)
         for i, text in enumerate(texts):
             # Use provided key otherwise use default key
-            key = keys[i] if keys else _redis_key(prefix)
+            key = keys[i] if keys else self._redis_key(prefix)
             metadata = metadatas[i] if metadatas else {}
-            pipeline.hset(
+            embedding = embeddings[i] if embeddings else self.embedding_function(text)
+            await pipeline.hset(
```

Comment:
I might be wrong, but I believe in async mode, `hset` doesn't have to be awaited everytime but only at `exceute`
https://redis-py.readthedocs.io/en/stable/connections.html#clusterpipeline-async